### PR TITLE
Add ASP.NET Core extension method guidance

### PR DESCRIPTION
**Pull request by pakrym**

### CODE REVIEW

File: docs/dotnet/implementation.md [hunk 1]
```diff
@@ -219,6 +219,55 @@ public readonly struct EncryptionAlgorithm : IEquatable<EncryptionAlgorithm>
 }
 ```
 
+## Client builder extensions
```
Comment (tg-msft) [inline docs/dotnet/implementation.md]: Maybe call it `ASP.NET Core Integration`?

File: docs/dotnet/implementation.md [hunk 2]
```diff
@@ -219,6 +219,55 @@ public readonly struct EncryptionAlgorithm : IEquatable<EncryptionAlgorithm>
 }
 ```
 
+## Client builder extensions
+
+All Azure client libraries ship with a set of extension methods that provide integration with ASP.NET Core applications by registering clients with DependencyInjection container, flowing Azure SDK logs to ASP.NET Core logging subsystem and providing ability to use configuration subsystem for client configuration (for more examplese see https://github.com/Azure/azure-sdk-for-net/tree/master/sdk/core/Microsoft.Extensions.Azure)
+
+{% include requirement/MUST id="dotnet-builder-class-name" %} provide a single `*ClientBuilderExtensions` class for every Azure SDK client library that contains client types. Name of the type should consist of last part of library name with `ClientBuilderExtensions` suffix added. For example: `SecretClientBuilderExtensions`, `BlobClientBuilderExtensions`
```
Comment (tg-msft) [inline docs/dotnet/implementation.md]: I wonder if taking `*ClientOptions` and changing to `*ClientBuilderExtensions` would be a simple naming rule to use here.

File: docs/dotnet/implementation.md [hunk 3]
```diff
@@ -219,6 +219,55 @@ public readonly struct EncryptionAlgorithm : IEquatable<EncryptionAlgorithm>
...
 
+## Client builder extensions
+
+All Azure client libraries ship with a set of extension methods that provide integration with ASP.NET Core applications by registering clients with DependencyInjection container, flowing Azure SDK logs to ASP.NET Core logging subsystem and providing ability to use configuration subsystem for client configuration (for more examplese see https://github.com/Azure/azure-sdk-for-net/tree/master/sdk/core/Microsoft.Extensions.Azure)
+
+{% include requirement/MUST id="dotnet-builder-class-name" %} provide a single `*ClientBuilderExtensions` class for every Azure SDK client library that contains client types. Name of the type should consist of last part of library name with `ClientBuilderExtensions` suffix added. For example: `SecretClientBuilderExtensions`, `BlobClientBuilderExtensions`
+
+{% include requirement/MUST id="dotnet-client-builder-class-namespace" %} use `Microsoft.Extensions.Azure` as a namespace.
+
+{% include requirement/MUST id="dotnet-client-builder-class-service-client" %} only provide overloads for the `*ServiceClient` if library ships one or for every client in assembly otherwise.
```
Comment (tg-msft) [inline docs/dotnet/implementation.md]: Change `overloads` to `integration extension methods` or similar.  Maybe rephrase this as something like `MUST provide integration extension methods for every top level client type users are expected to start with in the main namespace.  Do not include integration extension methods for secondary clients, child clients, or clients in advanced namespaces.`?

File: docs/dotnet/implementation.md [hunk 4]
```diff
@@ -219,6 +219,55 @@ public readonly struct EncryptionAlgorithm : IEquatable<EncryptionAlgorithm>
 }
 ```
 
+## Client builder extensions
+
+All Azure client libraries ship with a set of extension methods that provide integration with ASP.NET Core applications by registering clients with DependencyInjection container, flowing Azure SDK logs to ASP.NET Core logging subsystem and providing ability to use configuration subsystem for client configuration (for more examplese see https://github.com/Azure/azure-sdk-for-net/tree/master/sdk/core/Microsoft.Extensions.Azure)
```
Comment (tg-msft) [inline docs/dotnet/implementation.md]: Typo - examplese => examples

File: docs/dotnet/implementation.md (hunk 2 above)
Comment (maririos) [inline docs/dotnet/implementation.md]: Not sure if you mean that for every client under a client library a `*ClientBuilderExtensions` needs to exist

File: docs/dotnet/implementation.md [hunk 5]
```diff
@@ -219,6 +219,55 @@ public readonly struct EncryptionAlgorithm : IEquatable<EncryptionAlgorithm>
 }
 ```
 
+## Client builder extensions
+
+All Azure client libraries ship with a set of extension methods that provide integration with ASP.NET Core applications by registering clients with DependencyInjection container, flowing Azure SDK logs to ASP.NET Core logging subsystem and providing ability to use configuration subsystem for client configuration (for more examplese see https://github.com/Azure/azure-sdk-for-net/tree/master/sdk/core/Microsoft.Extensions.Azure)
+
+{% include requirement/MUST id="dotnet-builder-class-name" %} provide a single `*ClientBuilderExtensions` class for every Azure SDK client library that contains client types. Name of the type should consist of last part of library name with `ClientBuilderExtensions` suffix added. For example: `SecretClientBuilderExtensions`, `BlobClientBuilderExtensions`
+
+{% include requirement/MUST id="dotnet-client-builder-class-namespace" %} use `Microsoft.Extensions.Azure` as a namespace.
```
Comment (maririos) [inline docs/dotnet/implementation.md]: ohh so is this a new project under the same client library folder?

File: docs/dotnet/implementation.md (hunk 2 above)
Comment (pakrym) [reply docs/dotnet/implementation.md]: No, one per library. I'll try to improve wording.

File: docs/dotnet/implementation.md (hunk 5 above)
Comment (pakrym) [reply docs/dotnet/implementation.md]: Nope, same project just different namespace for this one file.

**Comment by maririos** [comment]
This might not be the place, but is there information on how to test or make sure that changes are good?

**Comment by pakrym** [comment]
> This might not be the place, but is there information on how to test or make sure that changes are good?

Not yet, I'm yet to write a good harness for testing this. Usually it "just works" because there is almost no custom code and you are calling into constructors covered by other tests but I agree there is a test gap here.

File: docs/dotnet/implementation.md [hunk 6]
```diff
@@ -219,6 +219,55 @@ public readonly struct EncryptionAlgorithm : IEquatable<EncryptionAlgorithm>
...
+
+``` C#
+public static IAzureClientBuilder<ConfigurationClient, ConfigurationClientOptions> AddConfigurationClient<TBuilder>(this TBuilder builder, string connectionString)
+    where TBuilder : IAzureClientFactoryBuilder
+{
+    return builder.RegisterClientFactory<ConfigurationClient, ConfigurationClientOptions>(options => new ConfigurationClient(connectionString, options));
+}
+```
+
+{% include requirement/MUST id="dotnet-client-builder-overload-tokencredential" %} provide extension method for `IAzureClientFactoryBuilderWithCredential` internface for constructors that take `TokenCredentials`. Extension method should take same set of parameters as constructor except the `TokenCredential` and call into `builder.RegisterClientFactory` overload that provides the token credential as part of factory lambda.
```
Comment (maririos) [inline docs/dotnet/implementation.md:248]: typo 
```suggestion
{% include requirement/MUST id="dotnet-client-builder-overload-tokencredential" %} provide extension method for `IAzureClientFactoryBuilderWithCredential` interface for constructors that take `TokenCredentials`. Extension method should take same set of parameters as constructor except the `TokenCredential` and call into `builder.RegisterClientFactory` overload that provides the token credential as part of factory lambda. 

```

File: docs/dotnet/implementation.md [hunk 7]
```diff
@@ -219,6 +219,55 @@ public readonly struct EncryptionAlgorithm : IEquatable<EncryptionAlgorithm>
...
+
+{% include requirement/MUST id="dotnet-client-builder-class-namespace" %} use `Microsoft.Extensions.Azure` as a namespace.
+
+{% include requirement/MUST id="dotnet-client-builder-class-service-client" %} provide integration extension methods for every top level client type users are expected to start with in the main namespace. Do not include integration extension methods for secondary clients, child clients, or clients in advanced namespaces.
+
+{% include requirement/MUST id="dotnet-client-builder-extension-name" %} name extension methods as `Add[ClientName]` for example. Add `AddSecretsClient`, `AddBlobServiceClient`.
+
+{% include requirement/MUST id="dotnet-client-builder-overloads" %} provide an overload for every set of constructor parameters.
+
+{% include requirement/MUST id="dotnet-client-builder-overload-normal" %} provide extension method for `IAzureClientFactoryBuilder` interface for constructors that take `TokenCredentials`. Extension method should take same set of parameters as constructor and call into `builder.RegisterClientFactory`
```
Comment (maririos) [inline docs/dotnet/implementation.md:236]: Not sure we actually need `TokenCredentials` here, as it seems that his overload will be for other auth methods different than `TokenCredentials`

**Comment by maririos** [comment]
I followed this documentation to create: https://github.com/Azure/azure-sdk-for-net/pull/10984

Thanks for putting it together @pakrym !

File: docs/dotnet/implementation.md (hunk 7 above)
Comment (pakrym) [reply docs/dotnet/implementation.md]: missed a "don't" 

`provide extension method for IAzureClientFactoryBuilder interface for constructors that don't take TokenCredentials.`
